File: back-go/domain/statistics/speedguards.go
package statistics

import (
	"math"

	"mystravastats/internal/shared/domain/strava"
)

const defaultPlausibleMaxSpeedMS = 35.0

func invalidSpeedSegmentPrefix(distances []float64, times []int, streamDataSize int, activityType string) []int {
	if streamDataSize < 2 {
		return make([]int, streamDataSize+1)
	}
	invalidSegments := make([]int, streamDataSize)
	threshold := plausibleSpeedThreshold(activityType)
	for i := 1; i < streamDataSize; i++ {
		invalidSegments[i] = invalidSegments[i-1]
		deltaTime := times[i] - times[i-1]
		deltaDistance := distances[i] - distances[i-1]
		if deltaTime <= 0 || deltaDistance < 0 {
			invalidSegments[i]++
			continue
		}
		speed := deltaDistance / float64(deltaTime)
		if speed > threshold {
			invalidSegments[i]++
		}
	}
	return invalidSegments
}

func hasInvalidSpeedSegment(invalidSegments []int, idxStart int, idxEnd int) bool {
	if idxStart < 0 || idxEnd <= idxStart || idxEnd >= len(invalidSegments) {
		return false
	}
	return invalidSegments[idxEnd] > invalidSegments[idxStart]
}

func IsPlausibleActivityMaxSpeed(activity *strava.Activity) bool {
	if activity == nil || math.IsNaN(activity.MaxSpeed) || math.IsInf(activity.MaxSpeed, 0) || activity.MaxSpeed <= 0 {
		return false
	}
	return activity.MaxSpeed <= plausibleSpeedThreshold(activity.Type)*1.3
}

func plausibleSpeedThreshold(activityType string) float64 {
	switch activityType {
	case "Run", "TrailRun":
		return 12
	case "Hike", "Walk":
		return 7
	case "AlpineSki":
		return 45
	default:
		return defaultPlausibleMaxSpeedMS
	}
}

File: back-kotlin/src/main/kotlin/me/nicolas/stravastats/domain/services/statistics/SpeedGuards.kt
package me.nicolas.stravastats.domain.services.statistics

import me.nicolas.stravastats.domain.business.strava.StravaActivity

private const val DEFAULT_PLAUSIBLE_MAX_SPEED_MS = 35.0

internal fun invalidSpeedSegmentPrefix(
    distances: List<Double>,
    times: List<Int>,
    streamDataSize: Int,
    activityType: String,
): IntArray {
    if (streamDataSize < 2) {
        return IntArray(streamDataSize + 1)
    }
    val invalidSegments = IntArray(streamDataSize)
    val threshold = plausibleSpeedThreshold(activityType)
    for (index in 1 until streamDataSize) {
        invalidSegments[index] = invalidSegments[index - 1]
        val deltaTime = times[index] - times[index - 1]
        val deltaDistance = distances[index] - distances[index - 1]
        if (deltaTime <= 0 || deltaDistance < 0.0 || deltaDistance / deltaTime > threshold) {
            invalidSegments[index]++
        }
    }
    return invalidSegments
}

internal fun hasInvalidSpeedSegment(invalidSegments: IntArray, idxStart: Int, idxEnd: Int): Boolean {
    if (idxStart < 0 || idxEnd <= idxStart || idxEnd >= invalidSegments.size) {
        return false
    }
    return invalidSegments[idxEnd] > invalidSegments[idxStart]
}

internal fun isPlausibleActivityMaxSpeed(activity: StravaActivity): Boolean {
    val maxSpeed = activity.maxSpeed.toDouble()
    return maxSpeed.isFinite() && maxSpeed > 0.0 && maxSpeed <= plausibleSpeedThreshold(activity.type) * 1.3
}

private fun plausibleSpeedThreshold(activityType: String): Double = when (activityType) {
    "Run", "TrailRun" -> 12.0
    "Hike", "Walk" -> 7.0
    "AlpineSki" -> 45.0
    else -> DEFAULT_PLAUSIBLE_MAX_SPEED_MS
}
